RLM_stable.thy

```diff
@@ -0,0 +1,1091 @@
+  then have "integrable  ?Dn1 (\<lambda> x. (\<Sum>i\<in>?I. ?l_swapped i x))" 
+    using integ_swap by auto
+  then have 4:"integrable  ?Dn1 (\<lambda> x. (\<Sum>i\<in>?I. ?l_swapped i x)/card ?I)"
+    by auto
+  have "(\<lambda> x. (\<Sum>i\<in>?I. ?l_swapped i x)/card ?I) =  (\<lambda> x. integral\<^sup>L ?pmfI (\<lambda> i. ?l_swapped i x))"
+  using  fin_I non_empty_I
+  by (simp add: integral_pmf_of_set[of ?I ]) 
+ 
+  then have 5:"integrable ?Dn1  (\<lambda> Sz. integral\<^sup>L ?pmfI (\<lambda> i. ?l_swapped i Sz) )" 
+    using 4 by auto
+
+   have "\<forall>S. \<forall>i\<in>?I. (truncated_S S n) i = S i"  using  truncated_S_def by auto 
+  then  have  "\<forall>S. integral\<^sup>L ?pmfI (\<lambda>i. ?l_trunc S (truncated_S S n i)) =
+               integral\<^sup>L ?pmfI (\<lambda>i. ?l_trunc S (S i))" 
+    by (metis (no_types, lifting) fin_I non_empty_I same_integral_restricted set_pmf_of_set)
+  then have 7:"(\<lambda> S. integral\<^sup>L ?pmfI (\<lambda>i. ?l_trunc S (truncated_S S n i))) =
+            (\<lambda> S.  integral\<^sup>L ?pmfI (\<lambda>i. ?l_trunc S (S i))) " by auto
+
+   have "n>0" using n_pos by auto
+  then have 6: "integrable ?Dn (\<lambda> S. integral\<^sup>L D (\<lambda> _.  integral\<^sup>L ?pmfI (?l_Si S))) =
+      integrable ?Dn1  (\<lambda> S.  integral\<^sup>L ?pmfI (\<lambda>i. ?l_trunc S (truncated_S S n i)))"
+    using 
+      integrable_uniform uniform_nn add_sample_integrable[of n " (\<lambda> S. (\<lambda> _.  
+       integral\<^sup>L ?pmfI (?l_Si S)))"]   by blast
+
+   have "card ?I = n" by auto
+  then have "\<forall> S. integral\<^sup>L ?pmfI (\<lambda>i. l (ridge_mine S k) (S i) ) =
+      (sum (?l_Si S) ?I) / card ?I"
+    using integral_pmf_of_set `finite ?I` `?I \<noteq> {}` by blast
+  then have "\<forall> S. train_err_loss S n l (ridge_mine S k) = 
+      integral\<^sup>L ?pmfI (?l_Si S)" 
+    using `card ?I = n` train_err_loss_def by force
+
+  then have 8:" integrable ?Dn  (\<lambda> S. train_err_loss S n l (ridge_mine S k)) 
+            =  integrable ?Dn1  (\<lambda> S.  integral\<^sup>L ?pmfI (?l_Si S))"
+    using 6 7 truncated_same_min  by auto
+
+  then have "integrable ?Dn1  (\<lambda> S.  integral\<^sup>L ?pmfI (?l_Si S))" using
+  train_err_integrable by auto
+
+  then have integrable_Si:"integrable (Samples (n+1) D) 
+      (\<lambda> S.  integral\<^sup>L (pmf_of_set {0..<n}) (\<lambda>i. l (ridge_mine S k) (S i)))"
+    by auto
+  have integrable_swapped: "integrable (Samples (n+1) D)
+                        (\<lambda> S.  measure_pmf.expectation (pmf_of_set {0..<n})
+                     (\<lambda> i. (l (ridge_mine (swapped_S S i n) k) (S i))))"
+    using 5 by auto
+  show ?thesis
+    using train_val_diff replace_one_stable2 integrable_D integrable_Si integrable_swapped 
+      pred_err_integrable  
+        train_err_integrable  by auto
+qed
+
+
+lemma oracle_inequality:
```

Comment:
this is the culmination of this theory, right? please document it and explain what it means, or where one can find the lemma in the informal text book